dali/operators/reader/loader/tar_utils.cc

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/tar_utils.h"
+#include <libtar.h>
+#include <algorithm>
+#include <cstdarg>
+#include <cstdlib>
+#include <string>
+#include <tuple>
+#include <utility>
+#include "dali/core/error_handling.h"
+
+namespace dali {
+namespace detail {
+
+constexpr uint64 kBlockSize = 512;
+constexpr uint64 kEmptyEofBlocks = 2;
+
+std::mutex instances_mutex;
+vector<TarArchive*> instances;
+
+int Register(TarArchive* archive) {
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  int instance_handle = -1;
+  for (auto& instances_entry : instances) {
+    if (instances_entry != nullptr) {
+      continue;
+    }
+    instances_entry = archive;
+    instance_handle = &instances_entry - instances.data();
+  }
+  if (instance_handle == -1) {
+    instance_handle = instances.size();
+    instances.push_back(archive);
+  }
+  return instance_handle;
+}
+
+TarArchive::TarArchive(std::unique_ptr<FileStream>&& stream)
+    : stream(std::forward<std::unique_ptr<FileStream>>(stream)),
+      eof(false),
+      instance_handle(Register(this)),
+      archiveoffset(0) {
+  this->stream->Seek(0);
+  ParseHeader();
+}
+
+TarArchive::TarArchive(TarArchive&& other) {
+  *this = std::forward<TarArchive>(other);
+}
+
+TarArchive& TarArchive::operator=(TarArchive&& other) {
+  stream = std::move(other.stream);
+  eof = other.eof;
+  filename = other.filename;
+  filesize = other.filesize;
+  readoffset = other.readoffset;
+  archiveoffset = other.archiveoffset;
+
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  instances[instance_handle] = nullptr;
+  instance_handle = other.instance_handle;
+  return *this;
+}
+
+TarArchive::~TarArchive() {
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  instances[instance_handle] = nullptr;
```

Comment:
I think that just for symmetry there should be an `Unregister` (inline) function.